presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestSimplifyPlanWithEmptyInput.java

```diff
@@ -0,0 +1,193 @@
+                                tableScan("orders", ImmutableMap.of("orderkey", "orderkey", "custkey", "custkey")))));
+    }
+
+    @Test
+    public void testUnionWithEmptyInput()
+    {
+        assertPlan("select orderkey, partkey from lineitem union all select orderkey, custkey as partkey from orders where false",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("orderkey", "partkey"),
+                        tableScan("lineitem", ImmutableMap.of("partkey", "partkey", "orderkey", "orderkey"))));
+    }
+
+    @Test
+    public void testSemiJoinEmptyFilterSource()
+    {
+        assertPlan("select orderkey, partkey from lineitem where orderkey in (select orderkey from orders where false)",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("orderkey", "partkey"),
+                        filter(
+                                "false",
+                                tableScan("lineitem", ImmutableMap.of("orderkey", "orderkey", "partkey", "partkey")))));
+    }
+
+    @Test
+    public void testSemiJoinEmptySource()
+    {
+        assertPlan("select orderkey, partkey from (select orderkey, partkey from lineitem where false) where orderkey in (select orderkey from orders)",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("orderkey", "partkey"),
+                        values("orderkey", "partkey")));
+    }
+
+    @Test
+    public void testAggregationWithDefaultOutput()
+    {
+        assertPlan("select count(*) as count from (select orderkey from orders where false)",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("count"),
+                        aggregation(
+                                ImmutableMap.of("count", functionCall("count", ImmutableList.of())),
+                                values())));
+    }
+
+    @Test
+    public void testAggregationNoDefaultOutput()
+    {
+        assertPlan("select orderkey, count(*) as count from (select orderkey from orders where false) group by orderkey",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("orderkey", "count"),
+                        values("orderkey", "count")));
+    }
+
+    @Test
+    public void testInnerJoinWithOverEmptyAggregation()
+    {
+        assertPlan("select o.orderkey, o.custkey, l.linenumber from orders o join (select orderkey, max(linenumber) as linenumber from lineitem where false group by orderkey) l on o.orderkey = l.orderkey",
+                enableOptimization(),
+                output(
+                        ImmutableList.of("orderkey", "custkey", "linenumber"),
+                        values("orderkey", "custkey", "linenumber")));
+    }
+}
```

Comment:
Also add some unit tests for simple nodes like limit/sort/filter/window. You can have it all in one sql query, where it all boils down to just empty values().

Reply:
Added